File: app/src/main/java/com/coinninja/coinkeeper/util/android/PreferencesUtil.java
package com.coinninja.coinkeeper.util.android;

import android.annotation.SuppressLint;
import android.content.SharedPreferences;
import android.os.AsyncTask;

import com.coinninja.coinkeeper.model.dto.AddressDTO;

import java.util.List;

import javax.inject.Inject;

public class PreferencesUtil {

    private final SharedPreferences preferences;
    PreferenceWriter preferenceWriter;
    private SharedPreferences.Editor editor;

    @SuppressLint("CommitPrefEdits")
    @Inject
    public PreferencesUtil(SharedPreferences preferences) {
        this.preferences = preferences;
        editor = preferences.edit();
        preferenceWriter = new PreferenceWriter();
    }

    public void savePreference(String key, long value) {
        editor.putLong(key, value);
        editor.apply();
    }

    public void savePreference(String key, boolean value, Callback callback) {
        preferenceWriter.clone()
                .setPreference(editor, callback, key, value)
                .execute();
    }

    public void savePreference(String key, boolean value) {
        if (editor == null) { return; }
        editor.putBoolean(key, value);
        editor.apply();
    }

    public void savePreference(String key, String value) {
        if (editor == null) { return; }
        editor.putString(key, value);
        editor.apply();
    }

    public void removePreference(String key) {
        if (editor == null) { return; }
        editor.remove(key);
        editor.apply();
    }

    public long getLong(String key, long value) {
        return preferences.getLong(key, value);
    }

    public boolean getBoolean(String key) {
        return getBoolean(key, false);
    }

    public boolean getBoolean(String key, boolean defaultValue) {
        return preferences.getBoolean(key, defaultValue);
    }

    public String getString(String key, String defaultValue) {
        return preferences.getString(key, defaultValue);
    }

    public boolean contains(String key) {
        return preferences.contains(key);
    }

    public interface Callback {
        void onComplete();
    }

    static class PreferenceWriter extends AsyncTask<Void, Void, Void> {

        Callback callback;
        SharedPreferences.Editor editor;

        @Override
        protected Void doInBackground(Void... voids) {
            editor.apply();
            return null;
        }

        @Override
        public PreferenceWriter clone() {
            return new PreferenceWriter();
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            super.onPostExecute(aVoid);
            callback.onComplete();
        }

        public PreferenceWriter setPreference(SharedPreferences.Editor editor, Callback callback, String key, boolean value) {
            this.editor = editor;
            this.callback = callback;
            this.editor.putBoolean(key, value);
            return this;
        }
    }
}

File: app/src/main/java/com/coinninja/coinkeeper/util/android/PreferencesUtil.kt
package com.coinninja.coinkeeper.util.android

import android.annotation.SuppressLint
import android.content.SharedPreferences
import android.os.AsyncTask
import app.dropbit.annotations.Mockable
import javax.inject.Inject

@Mockable
class PreferencesUtil @SuppressLint("CommitPrefEdits")
@Inject
constructor(internal val preferences: SharedPreferences) {
    internal var preferenceWriter: PreferenceWriter? = null
    internal val editor: SharedPreferences.Editor

    init {
        editor = preferences.edit()
        preferenceWriter = PreferenceWriter()
    }

    fun savePreference(key: String, value: Long) {
        editor.putLong(key, value)
        editor.apply()
    }

    fun savePreference(key: String, value: Boolean, callback: Callback) {
        preferenceWriter?.clone()?.setPreference(editor, callback, key, value)?.execute()
    }

    fun savePreference(key: String, value: Boolean) {
        editor.putBoolean(key, value)
        editor.apply()
    }

    fun savePreference(key: String, value: String) {
        editor.putString(key, value)
        editor.apply()
    }

    fun removePreference(key: String) {
        editor.remove(key)
        editor.apply()
    }

    fun getLong(key: String, value: Long): Long {
        return preferences.getLong(key, value)
    }

    @JvmOverloads
    fun getBoolean(key: String, defaultValue: Boolean = false): Boolean {
        return preferences.getBoolean(key, defaultValue)
    }

    fun getString(key: String, defaultValue: String): String? {
        return preferences.getString(key, defaultValue)
    }

    operator fun contains(key: String): Boolean {
        return preferences.contains(key)
    }

    fun removeAll() {
        editor.clear()
        editor.apply()
    }

    interface Callback {
        fun onComplete()
    }

    @Mockable
    internal class PreferenceWriter : AsyncTask<Void?, Void?, Void?>() {

        var callback: Callback? = null
        var editor: SharedPreferences.Editor? = null

        public override fun doInBackground(vararg voids: Void?): Void? {
            editor!!.apply()
            return null
        }

        fun clone(): PreferenceWriter {
            return PreferenceWriter()
        }

        public override fun onPostExecute(aVoid: Void?) {
            super.onPostExecute(aVoid)
            callback?.onComplete()
        }

        fun setPreference(editor: SharedPreferences.Editor?, callback: Callback, key: String, value: Boolean): PreferenceWriter {
            this.editor = editor
            this.callback = callback
            this.editor!!.putBoolean(key, value)
            return this
        }
    }
}
